File: TimberCruisingApp/ViewModels/ARBoundaryViewModel.swift
// Spec §5.1 ARBoundary + §7.8 + REQ-BND-001..004. Holds the plot center
// (set by a screen-space tap in Phase 3 — hit-tested to ground in iOS,
// placed at (0,0,0) in macOS previews), the fixed-area radius, and the
// live user pose so the screen can render a 72-vertex ring, classify
// stems as inside/outside/borderline, and raise the 15 m drift warn.
//
// The ground mesh is held as a `GroundMeshSnapshot` so slope correction
// (§7.8 step 2) is testable — iOS glue code refreshes the snapshot from
// ARMeshAnchor updates; macOS keeps it empty and the ring stays flat at
// the center's Y.

import Foundation
import Combine
import simd
import Common
import Models
import Sensors
import AR

@MainActor
public final class ARBoundaryViewModel: ObservableObject {

    // MARK: - Published surface

    /// World-space plot center. Nil until the user taps to set it.
    @Published public private(set) var centerWorld: SIMD3<Float>?

    /// 72 ring vertices + closure vertex, slope-corrected against the
    /// current GroundMeshSnapshot. Empty until center is set.
    @Published public private(set) var ringVertices: [SIMD3<Float>] = []

    /// Fixed-area plot radius. Default 11.28 m = √(400/π) → 1/25 ha.
    @Published public var radiusM: Float = 11.28

    /// Live XZ distance from the current camera to the plot center.
    @Published public private(set) var userDistanceM: Float = 0

    /// REQ-BND-004: true when user has walked > 15 m from center.
    @Published public private(set) var isDrifted: Bool = false

    /// Drift warn threshold in metres. Default 15 m.
    @Published public var driftRadiusM: Float = 15

    @Published public private(set) var trackingStatus: TrackingStatus = .notAvailable

    // MARK: - Dependencies

    public let session: ARKitSessionManager

    /// Latest ground mesh snapshot (world-space vertices + triangles).
    /// iOS-side glue refreshes this from ARMeshAnchor updates.
    public private(set) var groundMesh: GroundMeshSnapshot = .empty

    // MARK: - Construction

    public init(session: ARKitSessionManager? = nil) {
        self.session = session ?? ARKitSessionManager()
    }

    // MARK: - Lifecycle

    private var depthCancellable: AnyCancellable?
    private var trackingCancellable: AnyCancellable?

    public func onAppear() {
        session.run()
        trackingCancellable = session.$trackingStatus
            .sink { [weak self] status in
                self?.trackingStatus = status
            }
        depthCancellable = session.$latestDepthFrame
            .compactMap { $0 }
            .sink { [weak self] frame in
                let c = frame.cameraPoseWorld.columns.3
                self?.updateUserPosition(SIMD3<Float>(c.x, c.y, c.z))
            }
    }

    public func onDisappear() {
        session.pause()
        depthCancellable?.cancel(); depthCancellable = nil
        trackingCancellable?.cancel(); trackingCancellable = nil
    }

    // MARK: - Center placement

    /// Set the plot center at a world-space point. Typically the screen
    /// layer hit-tests a tap to the ground and hands the result here.
    public func setCenter(_ point: SIMD3<Float>) {
        centerWorld = point
        refreshRingVertices()
    }

    /// Production convenience: drop the center at the current camera's
    /// XZ projected onto the ground mesh (or the camera's own Y if no
    /// triangle contains the column). Returns false if no AR frame has
    /// been received yet.
    @discardableResult
    public func setCenterAtCurrentCamera() -> Bool {
        guard let frame = session.latestDepthFrame else { return false }
        let c = frame.cameraPoseWorld.columns.3
        let y = GroundMeshSampler.sampleGroundY(
            atX: c.x, z: c.z, snapshot: groundMesh) ?? c.y
        setCenter(SIMD3<Float>(c.x, y, c.z))
        return true
    }

    public func clearCenter() {
        centerWorld = nil
        ringVertices = []
        userDistanceM = 0
        isDrifted = false
    }

    /// Update the ground mesh (called by iOS glue from ARMeshAnchor
    /// updates). Re-runs slope correction on the existing ring.
    public func updateGroundMesh(_ snapshot: GroundMeshSnapshot) {
        groundMesh = snapshot
        refreshRingVertices()
    }

    // MARK: - Stem classification (REQ-BND-003)

    public func membership(forStemXZ stem: SIMD2<Float>) -> PlotBoundaryRenderer.StemMembership? {
        guard let c = centerWorld else { return nil }
        return PlotBoundaryRenderer.membership(
            stemPositionXZ: stem,
            centerXZ: SIMD2<Float>(c.x, c.z),
            radiusM: radiusM)
    }

    // MARK: - Internals

    private func refreshRingVertices() {
        guard let center = centerWorld else {
            ringVertices = []
            return
        }
        let flat = PlotBoundaryRenderer.ringVertices(
            center: center, radiusM: radiusM)
        ringVertices = PlotBoundaryRenderer.slopeCorrected(flat) { [groundMesh] x, z in
            GroundMeshSampler.sampleGroundY(atX: x, z: z, snapshot: groundMesh)
        }
    }

    /// Test hook — pump a camera position without an ARKit frame.
    public func updateUserPosition(_ point: SIMD3<Float>) {
        guard let c = centerWorld else {
            userDistanceM = 0
            isDrifted = false
            return
        }
        let dx = point.x - c.x
        let dz = point.z - c.z
        userDistanceM = sqrt(dx * dx + dz * dz)
        isDrifted = PlotBoundaryRenderer.isDriftedBeyond(
            userXZ: SIMD2<Float>(point.x, point.z),
            centerXZ: SIMD2<Float>(c.x, c.z),
            driftRadiusM: driftRadiusM)
    }

    // MARK: - Preview factory

    public static func preview(
        centerWorld: SIMD3<Float>? = nil,
        radiusM: Float = 11.28,
        userDistanceM: Float = 0,
        isDrifted: Bool = false
    ) -> ARBoundaryViewModel {
        let vm = ARBoundaryViewModel()
        vm.radiusM = radiusM
        if let c = centerWorld { vm.setCenter(c) }
        vm.userDistanceM = userDistanceM
        vm.isDrifted = isDrifted
        return vm
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/ui/screens/boundary/ARBoundaryViewModel.kt
// Port of iOS ViewModels/ARBoundaryViewModel.swift (spec §5.1 ARBoundary +
// §7.8 + REQ-BND-001..004). Holds the plot center (set from the current
// camera pose), the fixed-area radius, and the live user pose so the screen
// can render a 72-vertex ring, classify stems as inside/outside/borderline,
// and raise the 15 m drift warn.
//
// iOS keeps a `GroundMeshSnapshot` refreshed from ARMeshAnchor updates so
// slope correction is testable; ARCore has no scene-mesh anchor stream, so
// the sampler always returns null and the ring stays flat at the center's Y
// (the documented iOS degradation path).

package com.hcjeong.forestix.ui.screens.boundary

import com.hcjeong.forestix.ar.ArController
import com.hcjeong.forestix.ar.PlotBoundaryRenderer
import com.hcjeong.forestix.ar.Vec2
import com.hcjeong.forestix.ar.Vec3
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlin.math.sqrt

class ARBoundaryViewModel(
    /// Android analogue of the iOS ARKitSessionManager: the shared frame
    /// bridge the hosting ArCameraView keeps updated.
    val session: ArController,
) {

    // MARK: - Published surface

    /// World-space plot center. Null until the user taps to set it.
    private val _centerWorld = MutableStateFlow<Vec3?>(null)
    val centerWorld: StateFlow<Vec3?> = _centerWorld.asStateFlow()

    /// 72 ring vertices + closure vertex, slope-corrected against the
    /// ground sampler. Empty until center is set.
    private val _ringVertices = MutableStateFlow<List<Vec3>>(emptyList())
    val ringVertices: StateFlow<List<Vec3>> = _ringVertices.asStateFlow()

    /// Fixed-area plot radius. Default 11.28 m = sqrt(400/pi) -> 1/25 ha.
    private val _radiusM = MutableStateFlow(11.28f)
    val radiusM: StateFlow<Float> = _radiusM.asStateFlow()

    /// Live XZ distance from the current camera to the plot center.
    private val _userDistanceM = MutableStateFlow(0f)
    val userDistanceM: StateFlow<Float> = _userDistanceM.asStateFlow()

    /// REQ-BND-004: true when user has walked > 15 m from center.
    private val _isDrifted = MutableStateFlow(false)
    val isDrifted: StateFlow<Boolean> = _isDrifted.asStateFlow()

    /// Drift warn threshold in metres. Default 15 m.
    private val _driftRadiusM = MutableStateFlow(15f)
    val driftRadiusM: StateFlow<Float> = _driftRadiusM.asStateFlow()

    // MARK: - Lifecycle
    //
    // On iOS onAppear/onDisappear run and pause the ARKit session; on
    // Android the ArCameraView composable owns the ARCore session
    // lifecycle, so these are kept only for call-site parity.

    fun onAppear() {}

    fun onDisappear() {}

    // MARK: - Center placement

    /// Set the plot center at a world-space point.
    fun setCenter(point: Vec3) {
        _centerWorld.value = point
        refreshRingVertices()
    }

    /// Production convenience: drop the center at the current camera
    /// position. iOS additionally projects the camera XZ onto the LiDAR
    /// ground mesh; ARCore has no equivalent sampler, so this is the iOS
    /// fallback path (camera's own Y). Returns false if no AR frame has
    /// been received yet.
    fun setCenterAtCurrentCamera(): Boolean {
        val c = session.currentCameraPosition() ?: return false
        setCenter(c)
        return true
    }

    fun clearCenter() {
        _centerWorld.value = null
        _ringVertices.value = emptyList()
        _userDistanceM.value = 0f
        _isDrifted.value = false
    }

    fun setRadiusM(r: Float) {
        _radiusM.value = r
        refreshRingVertices()
    }

    // MARK: - Stem classification (REQ-BND-003)

    fun membership(forStemXZ: Vec2): PlotBoundaryRenderer.StemMembership? {
        val c = _centerWorld.value ?: return null
        return PlotBoundaryRenderer.membership(
            stemPositionXZ = forStemXZ,
            centerXZ = Vec2(c.x, c.z),
            radiusM = _radiusM.value,
        )
    }

    // MARK: - Internals

    private fun refreshRingVertices() {
        val center = _centerWorld.value
        if (center == null) {
            _ringVertices.value = emptyList()
            return
        }
        val flat = PlotBoundaryRenderer.ringVertices(
            center = center, radiusM = _radiusM.value)
        // No ground-mesh sampler on ARCore — flat ring (iOS sparse-mesh path).
        _ringVertices.value = PlotBoundaryRenderer.slopeCorrected(flat) { _, _ -> null }
    }

    /// Pump a camera position (polled from the AR frame by the screen).
    fun updateUserPosition(point: Vec3) {
        val c = _centerWorld.value
        if (c == null) {
            _userDistanceM.value = 0f
            _isDrifted.value = false
            return
        }
        val dx = point.x - c.x
        val dz = point.z - c.z
        _userDistanceM.value = sqrt(dx * dx + dz * dz)
        _isDrifted.value = PlotBoundaryRenderer.isDriftedBeyond(
            userXZ = Vec2(point.x, point.z),
            centerXZ = Vec2(c.x, c.z),
            driftRadiusM = _driftRadiusM.value,
        )
    }
}
